raven/utilities/forecasting.py

```diff
@@ -173,3 +182,79 @@ def perform_climatology_esp(model_name, forecast_date, forecast_duration, **kwds
+    Data comes from the CASPAR archive and must be aggregated such that each file
+    contains forecast data for a single day, but for all forecast timesteps and
+    all members.
+    
+    The code takes the region shapefile, the forecast date required, and the 
+    climate_model to use, here GEPS by default, but eventually could be 
+    GEPS, GDPS, REPS or RDPS.
+    '''
+
+    # Get the file locations and filenames as a function of the climate model and date
+    if climate_model is 'GEPS':
+        filein='https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/caspar/daily/GEPS_' + dt.datetime.strftime(date,'%Y%m%d') + '.nc'
+        
+        # Here we also extract the times at 6-hour intervals as Raven must have 
+        #constant timesteps.
+        times=[]
+        for single_time in (date + dt.timedelta(hours=n) for n in range(0,384,6)):
+            times.append(single_time)
+            
+    elif climate_model is 'GDPS':
+        filein='https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/caspar/daily/GDPS_' + dt.datetime.strftime(date,'%Y%m%d') + '.nc'
+        
+    elif climate_model is 'RDPS':
+        filein='https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/caspar/daily/REPS_' + dt.datetime.strftime(date,'%Y%m%d') + '.nc'
+        
+    elif climate_model is 'REPS':
+        filein='https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/caspar/daily/RDPS_' + dt.datetime.strftime(date,'%Y%m%d') + '.nc'
+        
+    # Extract the shapefile from the zipped package
+    tmp = Path(tempfile.mkdtemp())
+    ZipFile(region,'r').extractall(tmp)
+    shp = list(tmp.glob("*.shp"))[0]
+    vector = fiona.open(shp, "r")
+    shdf = [vector.next()["geometry"]]
+    
+    # extract the bounding box to subset the entire forecast grid to something
+    # more manageable
+    lon_min=vector.bounds[0]
+    lon_max=vector.bounds[2]
+    lat_min=vector.bounds[1]
+    lat_max=vector.bounds[3]
+
+    # Open the forecast data
+    ds=xr.open_dataset(filein)
+    
+    # Subset the data to the desired location (bounding box) and times
+    ds=subset.subset_bbox(ds, lon_bnds=[lon_min, lon_max], lat_bnds=[lat_min, lat_max]).sel(time=times)
+        
+    # Rioxarray requires CRS definitions for variables
+    # Here the name of the variable could differ based on the Caspar file processing
+    tas = ds.tas.rio.write_crs(4326)
+    pr = ds.pr.rio.write_crs(4326)
+    ds = xr.merge([tas, pr])
+            
+    # Now apply the mask of the basin contour and average the values to get a single time series
+    ds.rio.set_spatial_dims('rlon','rlat')
+
+    # rlon format is not changed by the subset.subset, do that here.
+    ds['rlon']=360-ds['rlon']
+    ds['rlon']=-ds['rlon']
+    
+    # clip the netcdf and average across space.
+    sub = ds.rio.clip(shdf, crs=4326)
+    sub = sub.mean(dim={'rlat','rlon'}, keep_attrs=True)
+     
+    return sub
```

Comment:
We should reuse the same functions used for the ECCC forecasts.

Reply:
@huard if you want I can try doing a first pass at these comments (and those of ECCC forecasts in the Geomet PR)? Or should I let @cjauvin give it a professional pass?

Reply:
It's really flattering to have the word "professional" used to describe me so often these days! :smile: But on the other hand I don't feel I particularly deserve it, it's just that I love Python and have been using it for many years!

Reply:
I mean... you are clearly and demonstrably light-years ahead of me in terms of Python coding, so from my point of view it's all the same! I can do the changes asked by @huard and you can clean up after if you want? that way you can focus longer on the routing toolbox?

My self-imposed goal is to have the Geomet/ECCC and hindcasting PRs merged in the next 10 days, so if that helps I'm ready to do it. Thanks!

Reply:
I think it's better if I take care of it, mainly because it's a good way for me to get a broader view of the code. I'll try to wrap it up quickly.

Reply:
OK cool, thanks! In any case, if there's anything that's unclear, just let me know!